sfge/src/main/java/com/salesforce/rules/PerformNullCheckOnSoqlVariables.java

```diff
@@ -0,0 +1,163 @@
+package com.salesforce.rules;
+
+import com.google.common.collect.ImmutableSet;
+import com.salesforce.config.UserFacingMessages;
+import com.salesforce.exception.ProgrammingException;
+import com.salesforce.exception.UnexpectedException;
+import com.salesforce.graph.ApexPath;
+import com.salesforce.graph.ops.expander.PathExpansionObserver;
+import com.salesforce.graph.source.ApexPathSource;
+import com.salesforce.graph.symbols.ScopeUtil;
+import com.salesforce.graph.symbols.SymbolProvider;
+import com.salesforce.graph.symbols.apex.ApexValue;
+import com.salesforce.graph.symbols.apex.Constraint;
+import com.salesforce.graph.vertex.BaseSFVertex;
+import com.salesforce.graph.vertex.SFVertex;
+import com.salesforce.graph.vertex.SoqlExpressionVertex;
+import com.salesforce.graph.vertex.VariableExpressionVertex;
+import java.util.*;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+
+public final class PerformNullCheckOnSoqlVariables extends AbstractPathTraversalRule
+        implements PathExpansionObserver {
+
+    private static final ImmutableSet<ApexPathSource.Type> SOURCE_TYPES =
+            ImmutableSet.copyOf(ApexPathSource.Type.values());
+
+    /**
+     * maps to true if a {@link com.salesforce.graph.vertex.VariableExpressionVertex} should cause a
+     * violation, false otherwise. If said vertex is not in the map, it should not get a violation
+     * either.
+     */
+    private final Map<VariableExpressionVertex, Boolean> violationInfos;
+
+    // use LazyHolder to get instance, but map needs to be instantiated
+    private PerformNullCheckOnSoqlVariables() {
+        this.violationInfos = new HashMap<>();
```

Comment:
Since rules are instantiated before thread-forking happens, I'm concerned this might be shared across threads.
Please try a case where there are multiple entrypoints (8 should be enough) with violations, to make sure you don't accidentally get duplicates.
If you do, then you'll probably need to move this into a separate class the way `RemoveUnusedMethod` has `UsageTracker`.

Reply:
Removed state!